parmed/openmm/parameters.py

```diff
@@ -166,6 +176,13 @@ def write(self, dest, provenance=None):
             if own_handle:
                 dest.close()
 
+    def _clean_params(self):
+        keep_types = set()
+        for name, residue in iteritems(self.residues):
+            for atom in residue.atoms:
+                keep_types.add(atom.type)
+        self.skip_types = [typ for typ in self.atom_types if typ not in keep_types]
```

Comment:
Ah, I see now that the cache is simply discarded and regenerated each time `write` is called.  OK, there are two ways of doing this:

1. Have a `prune` method that goes through and removes any types and parameters that have no representation in the residue templates.
2. Do what we're doing (prune at write time, but make sure they're still in the parameter set).

My initial preference was 1, since the intuitive behavior (to me) was for `write` to simply dump its entire database.  But I think I've convinced myself that what you're doing here is better.  It's simpler from an implementation perspective *and* more flexible from a practical perspective.

So all I think I'd like to see are better names.  This isn't "cleaning" the parameter set (since nothing is getting removed or deleted) -- it's simply finding unused types.  So what do you think about changing the `clean_params=False` keyword to `write_unused=True` (where `write_unused` corresponds to both atom types *and* parameters including those types)?  And per my suggestion above, what about changing `_clean_params` to `_find_unused_types`?

Also, the way that you're currently using `skip_types` -- which I agree is the correct way to do it -- should not be persistent (i.e., don't store it as an instance attribute -- generate it once inside `write` and pass it to all of the functions that need to use it).